lib/3scale/backend/worker_async.rb

```diff
@@ -74,21 +72,23 @@ def schedule_jobs
 
           break if @shutdown
 
-          @reactor.async { perform(job) }
-          scheduled += 1
-
-          break if @jobs.empty? or scheduled >= @max_concurrent_jobs
+          semaphore.async { perform(job) }
         end
+      ensure
+        barrier.wait
       end
 
-      def process_one
-        job = @job_fetcher.fetch
-        perform(job) if job
+      def clear_queue
+        barrier = Async::Barrier.new
+        semaphore = Async::Semaphore.new(@max_concurrent_jobs, parent: barrier)
+        semaphore.async { perform(@jobs.pop) } until @jobs.empty?
```

Comment:
Maybe use this, lest we produce a `ThreadError: queue empty`
```suggestion
        @jobs.length.times { semaphore.async { perform(@jobs.pop) } }
```

Example:
`@jobs` Queue contains 5 jobs and 5 async `perform` blocks have been registered. `@jobs.pop` runs in the background so 3 `performs` have been completed and 2 are to be completed. But at that time `@jobs.empty?` is still false so we schedule one or more `perform` blocks. That would leave us with for example 7 blocks resitered in total with 5 jobs in total. And 2 of the blocks will raise the `ThreadError: queue empty`.

I hope this makes sense.

Reply:
Also I think we need a test for the `clear_queue` method to make sure it cleans up all jobs from the queue as such algorithms are prone to errors. Although I'm more concerned a situation where for some reason we schedule more jobs in the `@jobs` Queue while `#clear_queue` is running. Although `fetch_jobs_thread.join` should take care of this, still a little scared from such a situation.

Reply:
> But at that time @jobs.empty? is still false so we schedule one or more perform blocks

After examining the code, I think that situation is theoretically possible.

>  Although I'm more concerned a situation where for some reason we schedule more jobs in the @jobs Queue while #clear_queue is running

We already fixed all inconsistencies we found so far that could lead to that situation: `process_all` runs under a `Sync` wrapper, and the children tasks are waited for with a barrier.

But we can't be 100% sure. If that happened then this wouldn't work:

`@jobs.length.times { semaphore.async { perform(@jobs.pop) } }`

Because we would create as much tasks as `@jobs.length` at the moment it was called, so if new jobs were added while `clear_queue` is running, those jobs would be lost.

I think the source of the problem is having `@jobs.pop` called from inside children tasks. So we can fix your both concerns by extracting the `pop` call to the parent task. I suggest something like this:

```ruby
def clear_queue
  barrier = Async::Barrier.new
  semaphore = Async::Semaphore.new(@max_concurrent_jobs, parent: barrier)
  
  until job = @jobs.pop
    semaphore.async { perform(job) }
  end
ensure
  barrier.wait
end
```

That way:

1. Tasks are only scheduled when there's a job for them
2. `@jobs.pop` will wait forever until the queue is closed, so theoretically no jobs could be lost.

WDTY?

Reply:
> Also I think we need a test for the clear_queue

That would take some time, I'll do it in the next few days.

Reply:
Yeah, I thought about it. I though implementing an algorithm longer than one line is not the ruby way and could be deprecated in the future. But lets do it anyway. I think it has to be
```
while job = @jobs.pop
```

Other than that - LGTM.

Reply:
I pushed some changes. I added a couple of tests for `#clear_queue` which if I'm right should cover all your worries. However, this is tricky and I could have missed something.